Modules/Core/SpatialObjects/include/itkImageMaskSpatialObject.hxx

```diff
@@ -25,292 +25,150 @@
+  IteratorType it( this->GetImage(),
+    this->GetImage()->GetLargestPossibleRegion() );
+  IteratorType prevIt( this->GetImage(),
+    this->GetImage()->GetLargestPossibleRegion() );
+  it.GoToBegin();
+  prevIt = it;
+
+  bool first = true;
+  PixelType outsideValue = NumericTraits< PixelType >::ZeroValue();
+  PixelType value = outsideValue;
+  PixelType prevValue = outsideValue;
+  IndexType tmpIndex;
+  PointType tmpPoint;
+  int count = 0;
+  int rowSize
+    = this->GetImage()->GetLargestPossibleRegion().GetSize()[0];
+  while ( !it.IsAtEnd() )
     {
-    index[i] = 0;
-    size[i] = 0;
-    }
-
-  if ( ImageType::ImageDimension == 3 )
-    {
-    for ( unsigned int axis = 0; axis < ImageType::ImageDimension; axis++ )
+    value = it.Get();
+    if ( value != prevValue || ( count == rowSize-1 && value != outsideValue ) )
       {
-      // Two slice iterators along each axis...
-      // Find the orthogonal planes for the slices
-      unsigned int i, j;
-      unsigned int direction[2];
-      for ( i = 0, j = 0; i < 3; ++i )
+      prevValue = value;
+      if( value == outsideValue )
         {
-        if ( i != axis )
-          {
-          direction[j] = i;
-          j++;
-          }
+        tmpIndex = prevIt.GetIndex();
         }
-
-      // Create the forward iterator to find lower bound
-      SliceIteratorType fit( image,  image->GetRequestedRegion() );
-      fit.SetFirstDirection(direction[1]);
-      fit.SetSecondDirection(direction[0]);
-
-      fit.GoToBegin();
-      while ( !fit.IsAtEnd() )
+      else
         {
-        while ( !fit.IsAtEndOfSlice() )
-          {
-          while ( !fit.IsAtEndOfLine() )
-            {
-            if ( fit.Get() != outsideValue )
-              {
-              index[axis] = fit.GetIndex()[axis];
-              fit.GoToReverseBegin(); // skip to the end
-              break;
-              }
-            ++fit;
-            }
-          fit.NextLine();
-          }
-        fit.NextSlice();
+        tmpIndex = it.GetIndex();
         }
-
-      // Create the reverse iterator to find upper bound
-      SliceIteratorType rit( image,  image->GetRequestedRegion() );
-      rit.SetFirstDirection(direction[1]);
-      rit.SetSecondDirection(direction[0]);
-
-      rit.GoToReverseBegin();
-      while ( !rit.IsAtReverseEnd() )
+      this->GetImage()->TransformIndexToPhysicalPoint( tmpIndex, tmpPoint );
+      if( first )
         {
-        while ( !rit.IsAtReverseEndOfSlice() )
-          {
-          while ( !rit.IsAtReverseEndOfLine() )
-            {
-            if ( rit.Get() != outsideValue )
-              {
-              size[axis] = rit.GetIndex()[axis] - index[axis] + 1;
-              rit.GoToBegin(); //Skip to reverse end
-              break;
-              }
-            --rit;
-            }
-          rit.PreviousLine();
-          }
-        rit.PreviousSlice();
+        first = false;
+        this->GetMyBoundingBoxInObjectSpace()->SetMinimum( tmpPoint );
```

Comment:
Related to the other comment, if this is already valid, the const_cast can be removed?

Reply:
This shouldn't be valid.  GetMyBoundingBoxInObjectSpace is declared as const, but was returning a non-const pointer to the member variable.   Fixed GetMyBoundingBoxInObjectSpace to return a const pointer (and be a const function) and added the GetModifiableMyBoundingBoxInObjectSpace as a protected member function.